### PR TITLE
feat: filter autocomplete

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #1660 

### Description:

Add filter autocomplete:
- Autocomplete only available for string dimensions since it covers the majority of the use cases. We might extend this feature to other types later on.
- Autocomplete available in explore page and dashboard page
- Each search fetches 10 results 
- We don't cache values in our Database

filter autocomple 2

### CODE REVIEW

**Comment by TuringLovesDeathMetal** [comment]
(forgot to say in my last comment: THIS IS AMAZING!!!! 🚀 )

**Comment by TuringLovesDeathMetal** [comment]
> Each search fetches 5 results ( open to feedback )

I think that we should return all of the results. The use case for this is mostly when I'm using the autocomplete on an empty filter to see what values are available. (i.e. the suggestion I posted above). 

For example, if I'm a business user, and I want to be able to filter on the different user types in a dashboard, I might not actually know how the data team have named these user types. 

Is it `lightdash cloud`, `cloud`, `paid`, `US instances`?

If we only match the first 5 results, then you might not see the option that you're looking for. 

So now, you'll have to click on `user type` in the results table, run the query, and find the value you're looking for (basically, the same functionality as we have now).

**Comment by ZeRego** [comment]
> I think that we should return all of the results. 

@TuringLovesDeathMetal What if there are hundreds or thousands of results ?
Providing a small subset of results seems to be the norm when searching in other apps. e.g slack search 

> For example, if I'm a business user, and I want to be able to filter on the different user types in a dashboard, I might not actually know how the data team have named these user types.
> Is it lightdash cloud, cloud, paid, US instances?
> If we only match the first 5 results, then you might not see the option that you're looking for.

From my experience using slack search I try each word until I find what I want. 

Some have a button to see all the results like our docs website but I rarely end up using.
Screenshot 2022-09-12 at 13 42 12

**Comment by TuringLovesDeathMetal** [comment]
cool, ya let's add the first 5 results then :) and if someone complains, we can update (and make it like the first 20 or something 😅 )

Could we maybe boost it to 10? I don't know why, but it just feels better haha

File: packages/backend/src/routers/projectRouter.ts [hunk 1]
```diff
@@ -183,6 +183,37 @@ projectRouter.post(
...
+    isAuthenticated,
+    async (req, res, next) => {
+        try {
+            const value: string =
+                typeof req.query.value === 'string'
+                    ? req.query.value.toString()
+                    : '';
+            const limit: number =
+                typeof req.query.limit === 'string'
+                    ? parseInt(req.query.limit.toString(), 10)
```
Comment (rephus) [inline packages/backend/src/routers/projectRouter.ts:197]: why do you need to do this verbosy type checking ?

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 2]
```diff
@@ -578,6 +583,87 @@ export class ProjectService {
...
+
+        const distinctMetric: AdditionalMetric = {
+            name: `${field.name}_distinct`,
+            label: `Distinct of ${field.label}`,
+            table: field.table,
+            sql: `DISTINCT ${field.sql}`,
+            type: MetricType.STRING,
+        };
+
+        const metricQuery: MetricQuery = {
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:625]: :exploding_head:

File: packages/frontend/src/components/common/Filters/FilterInputs/StringAutoComplete.tsx [hunk 3]
```diff
@@ -27,14 +34,68 @@ function itemPredicate(
...
 
-const StringMultiSelect: FC<Props> = ({ values, suggestions, onChange }) => {
+export const useDebouncedSearch = (
+    projectUuid: string,
+    fieldId: string,
+    query: string | undefined,
+    enabled: boolean,
+) => {
+    const [debouncedQuery, setDebouncedQuery] = useState<string>();
+    useDebounce(
```
Comment (rephus) [inline packages/frontend/src/components/common/Filters/FilterInputs/StringAutoComplete.tsx:50]: :ok_hand:

**Review by rephus — APPROVED** [review]
Clever implementation :+1:  And it works nicely.

File: packages/backend/src/routers/projectRouter.ts (hunk 1 above)
Comment (ZeRego) [reply packages/backend/src/routers/projectRouter.ts:197]: what part ? checking the type, toString, parseInt
Just based myself on existing code from another endpoint with url params. 
The `toString` might be unnecessary.

**Pull request merged**